R/tfpscanner.R

```diff
@@ -62,698 +70,796 @@ tfpscan <- function(tre,
+  { # nolint
+    n <- ape::Ntip(tre)
+    nnode <- ape::Nnode(tre)
+    poedges <- tre$edge[ape::postorder(tre), ]
+    preedges <- tre$edge[rev(ape::postorder(tre)), ]
+
+
+    # PRECOMPUTE for each node
+    # num descendents
+    ndesc <- rep(0, n + nnode) # n tips descending
+    ndesc[1:n] <- 1
+    Ndesc <- rep(1, n + nnode) # include internal nodes
+    # max and min sample times of descendants :
+    tlsts <- sts[tre$tip.label]
+    max_desc_time <- rep(-Inf, n + nnode)
+    max_desc_time[1:n] <- tlsts
+    min_desc_time <- rep(Inf, n + nnode)
+    min_desc_time[1:n] <- tlsts
+    # postorder traverse
+    for (ie in seq_len(nrow(poedges))) {
+      a <- poedges[ie, 1]
+      u <- poedges[ie, 2]
+      ndesc[a] <- ndesc[a] + ndesc[u]
+      Ndesc[a] <- Ndesc[a] + Ndesc[u]
+      max_desc_time[a] <- max(max_desc_time[a], max_desc_time[u])
+      min_desc_time[a] <- min(min_desc_time[a], min_desc_time[u])
+    }
+    clade_age <- max_desc_time - min_desc_time # time span of descendant tips
+
+
+    # descendants; note also counts mrca node
+    descendants <- lapply(1:(n + nnode), function(u) integer(Ndesc[u])) # pre allocate
+    for (u in 1:(n + nnode)) {
+      descendants[[u]][1] <- u
+    }
+    Ndesc_index <- rep(2, n + nnode)
+    for (ie in seq_len(nrow(poedges))) {
+      a <- poedges[ie, 1]
+      u <- poedges[ie, 2]
+      ## mem efficient way to fill in values for a
+      i0 <- Ndesc_index[a]
+      i1 <- Ndesc[u] + i0 - 1
+      descendants[[a]][i0:i1] <- descendants[[u]]
+      Ndesc_index[a] <- i1 + 1
+    }
+
+    # descendant tips; index of tip
+    descendantTips <- descendants
+    for (a in (n + 1):(n + nnode)) {
+      descendantTips[[a]] <- descendantTips[[a]][descendantTips[[a]] <= n]
+    }
+    ## tip labels, only including names in amd which meet inclusion criteria (excl NA)
+    descendantSids <- lapply(
+      1:(n + nnode),
+      function(u) stats::na.omit(tre$tip.label[descendantTips[[u]]])
+    )
+
+    # ancestors
+    st0 <- Sys.time()
+    ancestors <- lapply(1:(n + nnode), function(u) integer())
+    for (ie in seq_len(nrow(preedges))) {
+      a <- preedges[ie, 1]
+      u <- preedges[ie, 2]
+      ancestors[[u]] <- c(ancestors[[a]], a)
+    }
+    st1 <- Sys.time()
+
+    ## vector samp time desc
+    descsts <- NULL # deprecate
+  }
+
+  message(paste("Derived lookup variables", Sys.time()))
+
+  .get_comparator_ancestor <- function(u,
+                                       num_comparison = num_ancestor_comparison) {
+    nu <- ndesc[u]
+    asu <- ancestors[[u]]
+    na <- -1
+    for (a in asu) {
+      na <- ndesc[a]
+      nc <- na - nu
+      if (nc >= num_comparison) {
+        break
+      }
+    }
+    if (na < nu) {
+      if (na > 0) {
+        message("Failed to find comparator ancestor with more tips. Returning NA. Node: ", u)
+      }
+      return(NA)
+    }
+    a
+  }
+
+  # matched by time and in proportion to region prevalence
+  .get_comparator_sample <- function(u,
```

Comment:
This function was extensively refactored in mrc-ide. The only difference between this and their version is that stats::na.omit is used and the function params are split over two lines. You can git diff against a08bbd6 to see differences against a styled version of the mrc-ide master

Reply:
Checked this commit and everything good